src/daemon/http/tls.rs

```diff
@@ -1,322 +1,380 @@
-}
+//------------ create_server_config -----------------------------------------
 
-impl<T: AsyncWrite + Unpin> AsyncWrite for LiftIo<T> {
-    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<io::Result<usize>> {
-        Pin::new(&mut self.get_mut().0).poll_write(cx, buf)
-    }
+/// Creates the TLS server config.
+pub fn create_server_config(
+    key_path: &Path, cert_path: &Path
+) -> Result<ServerConfig, TlsConfigError> {
+    let mut config = ServerConfig::builder()
+        .with_no_client_auth()
+        .with_single_cert(
+            read_certs(cert_path)?, read_key(key_path)?
+        ).map_err(|err| TlsConfigError::other(ErrorKind::Tls, err))?;
 
-    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        Pin::new(&mut self.get_mut().0).poll_flush(cx)
+    if std::env::var(SSLKEYLOGFILE_ENV_VAR_NAME).is_ok() {
+        config.key_log = Arc::new(KeyLogFile::new());
     }
 
-    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        Pin::new(&mut self.get_mut().0).poll_shutdown(cx)
-    }
+    Ok(config)
 }
 
-impl<T: AsyncRead + AsyncWrite + Unpin> Transport for LiftIo<T> {
-    fn remote_addr(&self) -> Option<SocketAddr> {
-        None
-    }
+/// Reads the certificates from the given PEM file.
+fn read_certs(
+    path: &Path
+) -> Result<Vec<CertificateDer<'static>>,TlsConfigError> {
+    rustls_pemfile::certs(
+        &mut io::BufReader::new(
+            File::open(path).map_err(|err| {
+                TlsConfigError::new(ErrorKind::Cert(path.into()), err)
+            })?
+        )
+    ).collect::<Result<_, _>>().map_err(|err| {
+        TlsConfigError::new(ErrorKind::Cert(path.into()), err)
+    })
 }
 
-/// Represents errors that can occur building the TlsConfig
-#[derive(Debug)]
-pub(crate) enum TlsConfigError {
-    Io(io::Error),
-    /// An Error parsing a Pkcs8 key
-    Pkcs8ParseError,
-    /// An Error parsing a Rsa key
-    RsaParseError,
-    /// An error from an empty key
-    EmptyKey,
-    // /// An error from an invalid key
-    // InvalidKey(TLSError),
-    Rustls(tokio_rustls::rustls::Error),
-}
-
-impl std::fmt::Display for TlsConfigError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            TlsConfigError::Io(err) => err.fmt(f),
-            TlsConfigError::Pkcs8ParseError => write!(f, "pkcs8 parse error"),
-            TlsConfigError::RsaParseError => write!(f, "rsa parse error"),
-            TlsConfigError::EmptyKey => write!(f, "key contains no private key"),
-            TlsConfigError::Rustls(err) => write!(f, "rustls error: {}", err),
+/// Reads the first private key from the given PEM file.
+///
+/// The key may be a PKCS#1 RSA private key, a PKCS#8 private key, or a
+/// SEC1 encoded EC private key. All other PEM items are ignored.
+///
+/// Errors out if opening or reading the file fails or if there isn’t exactly
```

Comment:
The while loop seems to try to the next item if the current one is not supported... so this comment is incorrect I think?

Reply:
Changed it to “… if there isn’t exactly one _recognized_ private key in the file.”